src/validation.cpp

```diff
@@ -2927,13 +2927,24 @@ bool CChainState::ActivateBestChain(BlockValidationState& state, const CChainPar
                         assert(blockindex);
 
                         auto tmp_set = setBlockIndexCandidates;
+                        bool canBeConnected = false;
                         for (auto* candidate : tmp_set) {
+                            bool hasTxsAndValid = candidate->HaveTxsDownloaded() && TestBlockIndex(candidate);
                             // if candidate has txs downloaded & currently arrived block is ancestor of `candidate`
-                            if (candidate->HaveTxsDownloaded() && TestBlockIndex(candidate) &&
-                                candidate->GetAncestor(blockindex->nHeight) == blockindex) {
+                            if (hasTxsAndValid && candidate->GetAncestor(blockindex->nHeight) == blockindex) {
                                 // then do pop fr with candidate, instead of blockindex
                                 pindexBestChain = VeriBlock::compareTipToBlock(candidate);
+                                canBeConnected = true;
                             }
+
+                            if (hasTxsAndValid && blockindex->GetAncestor(candidate->nHeight) == candidate) {
+                                canBeConnected = true;
+                            }
+                        }
+
+                        // do not try connecting block if there are no candidates
+                        if (!canBeConnected) {
+                            break;
```

Comment:
hmmm. in original Bitcoin algorithm is to set "pindexBestChain = nullptr" if there's no candidates. 
Changing this may break invariants.

Reply:
But setting "pindexBestChain = nullptr" forces FindBestChain on all candidates, which we want to avoid. What if we try to add this bypass to the original bitcoin algorithm (without isPopActive() check) and look if it breaks any tests?

Reply:
okay, I reviewed more, it should be fine to break

Reply:
I've tried to move a bypass to the original bitcoin code, so it works on all blocks disregard POP protocol activation. All tests pass. So we can even consider pushing this fix to the upstream.
You can approve the fix if all set.